src/pkg/podman/podman.go

```diff
@@ -226,18 +237,80 @@ func IsToolboxImage(image string) (bool, error) {
 	return true, nil
 }
 
-// Pull pulls an image
-func Pull(imageName string) error {
+func Login(registry string, extraArgs ...string) error {
+	var stderr bytes.Buffer
+
 	logLevelString := LogLevel.String()
-	args := []string{"--log-level", logLevelString, "pull", imageName}
+	args := []string{"--log-level", logLevelString, "login", registry}
+	args = append(args, extraArgs...)
 
-	if err := shell.Run("podman", nil, nil, nil, args...); err != nil {
+	if err := shell.Run("podman", os.Stdin, os.Stdout, &stderr, args...); err != nil {
+		if stderr.Len() == 0 {
+			return fmt.Errorf("failed to login to %s: %w", registry, err)
+		}
+
+		err := parseErrorMsg(&stderr).(*internalError)
 		return err
 	}
 
 	return nil
 }
 
+// Pull pulls an image. Wraps around command 'podman pull'.
+//
+// The image pull can fail for many reasons. The recognized reasons are:
+//   - manifest unknown (ErrManifestUnknown)
+//   - name unknown (ErrNameUnknown)
+//   - unauthorized (ErrUnauthorized)
+//
+// To learn more about OCI API error codes, see: https://github.com/opencontainers/distribution-spec/blob/main/spec.md
+//
+// The wrapper tries to discern these errors to some extent and provide means
+// for their handling. The provided errors can not be relied on completely as
+// they are created based on error message parsing that differs based on used
+// registry.
+func Pull(imageName string) error {
+	var stderr bytes.Buffer
+
+	logrus.Debugf("Pulling image %s", imageName)
+
+	logLevelString := LogLevel.String()
+	args := []string{"--log-level", logLevelString, "pull", imageName, "--quiet"}
```

Comment:
Umm... why `--quiet`? We do expose the entire spew from `podman pull ...` when `--verbose` is enabled.

Reply:
Oh, because the progress information from `podman pull` interferes with the screen scraping. :(

This is one reason why screen scraping is horrible. We need to find a way to do this without using `--quiet` because having the actual progress bar show up when using `--verbose` is quite useful for debugging. eg., you get to see if things are just slow or stuck.

Reply:
The progress bar won't show up regardless of the use of `--quiet` since `stderr` is redirected into a bytes buffer.

Is there anything else we can do instead of screen scraping? We could either print the content of the bytes buffer once the execution of Pull is done or somehow pass it through to Toolbox's `stderr` when debug logging is enabled. This will still interfere with the parsing though.

One way around could be to update the error parsing method to start parsing upon reading `Error:` in the output. That usually signals the presence of an error.

Reply:
I have a potential solution in #823.

Reply:
I tried to address this issue. Try it.